### ISSUE TITLE
Device-Head-Transformation not updated in FieldTrip after applying Maxfilter with head position transformation in MNE

### ISSUE THREAD

**Issue by tillhabersetzer** [fieldtrip/fieldtrip#2350]
**Goal**
I am analyzing MEG data obtained from a Neuromag TRIUX, consisting of multiple runs with slight variations in head positions. My aim is to merge these runs with matching head positions, allowing me to perform the coregistration between the head coordinate system and MRI just once.

Two toolboxes are used during the analysis, FieldTrip and MNE. Since I've identified an issue within FieldTrip, I'll begin by addressing it here.
Windows 10, Matlab R2023a
fieldtrip-20231113   
MNE version 1.6

**Description**
I applied MNE's Maxwell filter (https://mne.tools/stable/generated/mne.preprocessing.maxwell_filter.html), specifying a distinct head position by using one of the recorded runs as a reference destination head position. 
The MEG device-head-transformation matrix in MNE raw_tsss.info['dev_head_t'] showed the expected change. 
The transformation now matches that of the reference file.
Maxfiltered data imported into FieldTrip displayed no change in the MEG device-to-head-transformation matrix or sensor locations within the grad structure, which should align with the transformed head coordinate system.
Comparing a maxfiltered file with and without head position transformation in FieldTrip revealed differences in the recorded data (data.trial{1}), despite unchanged transformation and sensor details.
Manually altering the device-to-head-transformation matrix in FieldTrip and resaving it as a fif file reflected the anticipated changes in sensor locations within the grad structure.
The issue seems to be with FieldTrip not correctly reading the device-to-head transformation after the data has been repositioned using maxfilter.

**Reproduce**
1.)
Download data and example scripts
https://cloudstorage.elearning.uni-oldenburg.de/s/3mii9Ws6rKsRaye
password: DcDojcktS7

2.) Maxfilter data (Python) - maxfilter.py

import mne

raw1 = mne.io.read_raw_fif('sub-05_run-1.fif', allow_maxshield=False, verbose=True)
dev_head_t1 = raw1.info['dev_head_t']
raw2 = mne.io.read_raw_fif('sub-05_run-2.fif', allow_maxshield=False, verbose=True)
dev_head_t2 = raw2.info['dev_head_t']

% Check that dev_head_t1 and dev_head_t2 are different 

% Apply MaxFilter - with headposition transformation
raw1_tsss = mne.preprocessing.maxwell_filter(
    raw1, cross_talk='ct_sparse.fif', calibration='sss_cal.dat', 
    st_duration=10, head_pos=None, destination=dev_head_t2, coord_frame='head', verbose=True)     
dev_head_tsss = raw1_tsss.info['dev_head_t']
% Check that dev_head_t2 and dev_head_tsss are same

% save
raw1_tsss.save('sub-05_run-1_tsss.fif',overwrite=True)

3.) Load data into FieldTrip (Matlab) - inspect_data.m

close all
clear all
clc

% Load different headers and compare
%--------------------------------------------------------------------------
hdr1      = ft_read_header('sub-05_run-1.fif');
hdr2      = ft_read_header('sub-05_run-2.fif');
hdr1_tsss = ft_read_header('sub-05_run-1_tsss.fif');

% Compare transformations
%------------------------
hdr1.orig.dev_head_t.trans
hdr2.orig.dev_head_t.trans
hdr1_tsss.orig.dev_head_t.trans
% No change in trafo hdr1 and hdr1_tsss -> No effect of Maxfilter
isequal(hdr1.orig.dev_head_t.trans,hdr1_tsss.orig.dev_head_t.trans)

% Same holds for sensor / channel positions
%------------------------------------------
hdr1.grad.chanpos
hdr2.grad.chanpos
hdr1_tsss.grad.chanpos
isequal(hdr1.grad.chanpos,hdr1_tsss.grad.chanpos) % minor numerical differences but still the same

% Workaround
%--------------------------------------------------------------------------
% Change transformation matrix manually
cfg             = [];
cfg.dataset     = 'sub-05_run-1.fif';
data1_corrected = ft_preprocessing(cfg);

data1_corrected.hdr.orig.dev_head_t.trans = hdr2.orig.dev_head_t.trans; % use run2 as reference
fiff_file = 'sub-05_run-1_hp_corrected.fif';
fieldtrip2fiff(fiff_file, data1_corrected)

% Check headposition
hdr1_corrected = ft_read_header('sub-05_run-1_hp_corrected.fif');
isequal(hdr2.orig.dev_head_t.trans,hdr1_corrected.orig.dev_head_t.trans)
% Same tranformation for run1 and run2

% Same holds for sensor positions
% same after manual correction (numerical issues)
isequal(hdr2.grad.chanpos,hdr1_corrected.grad.chanpos)

**Comment by schoffelen** [comment]
Hi @tillhabersetzer, thanks for reporting. I am aware that there are suboptimalities w.r.t. the handling of sensor coil coordinates in the context of maxfiltered data. Since the core developers of FieldTrip do not work routinely with Elekta data, we have never made this a big priority, nor are experts on the matter (although we have been learning about this along the way). Would you be willing to take the challenge, and implement fieldtrip support for the correct reading of the sensor positions? (that is: without the workaround that you suggest?). Also - for that matter - would you be able to also upload a version of the maxfiltered data, the output of your step 2)? The latter would save us a little bit of hassle.

I think that the key to fixing this is to make some appropriate changes in the lower level mne2grad function, which converts the fif-file's chaninfo structure into a representation that fieldtrip understands. I would assume that - in case of tsss'ed data with head position correction, this requires the application of an extra transformation matrix to the coil positions, but perhaps also an optional storage of the spherical harmonics expansion that has been used to transform the data (although we'd perhaps need to think through the latter).

**Comment by tillhabersetzer** [comment]
Hi @schoffelen, thanks for the lightning-fast response. Yes, I'm eager to take on the challenge and hopefully solve it.

I started to dive into the problem by checking what's going on in mne2grad function as you suggested.
I tried to read the header from the different files (no maxfilter, maxfilter with/ without head trafo apllied) and I noticed that mne2grad already got the wrong (old) dev-head-trafo as input (line 1919 in ft_read_header.m).
The problem seems to be that fiff_read_meas_info(filename) returns the wrong dev-head-trafo (line 1910 in ft_read_header.m).

In fiff_read_meas_info the variable dev_head_t is retrieved from the fiff file multiple times (line 144, line 202).
I noticed:
A non-maxfiltered file retrieves the dev-head-trafo only once (line 202)
A maxfiltered file retrieves the dev-head-trafo twice (line 144 and line 202)

If the head position has not changed for the maxfiltered file, both dev-head-trafos (l. 144, l. 202) are the same.
But if the head position has changed, the dev-head-trafo in l. 144 is the new one which is afterwards overwritten by the old one in l. 202. 

I dont really know why the dev-head-trafo is picked up twice in case the file has been maxfiltered. Maybe you know more about it? 
But l.202 is always executed cause of l. 191 (isempty(dev_head_t) || isempty(ctf_head_t)). There is no ctf_head_t in the neuromag file, so it is always overwritten. 

I checked all this by calling these lines in my uploaded script (inspect_data.m)
hdr1           = ft_read_header('sub-05_run-1.fif');
hdr2           = ft_read_header('sub-05_run-2.fif');
hdr1_tsss      = ft_read_header('sub-05_run-1_tsss.fif');
hdr1_tsss_nohp = ft_read_header('sub-05_run-1_tsss_nohp.fif');

**Comment by larsoner** [comment]
> The problem there is, that maintaining the matlab code understandably has relatively low priority for said MNE folks. However, I have already discussed some potential improvement of the mne-matlab code (for other functions) with @larsoner, but in order to take that to the next level it was agreed that I should first implement a unit test framework to benchmark the current state of the (somewhat outdated) code. I have made some small steps into that direction recently, but it's not finished yet.

I don't think that needs to stop us from making bugfixes when needed. +1 for someone opening a PR to `mne-matlab` to fix the bug, then FieldTrip can just re-vendor the updated file. I think this is a better pattern to follow than fixes landing in FieldTrip directly and updating mne-matlab later.

Of course if it's easier for you in a single PR to FieldTrip update the files and some tests or whatever, see that it works, then open a PR to mne-matlab with the changes that's fine. But it would be good to merge the mne-matlab PR *before* the one to FieldTrip. Otherwise we risk the "later point" in pushing things to the rest of the world becoming "never" :)
